### ISSUE TITLE
Intellisense doesn't work on my solution

### ISSUE THREAD

**Issue by frblondin** [dotnet/roslyn#21629]
**Version Used**: VS2017 15.3.1

**Steps to Reproduce**:
Note that these steps are only reproductible on a particular solution. The same steps are not failing on other solutions
1. Add a new Class1.cs type, make it implementing IDisposable (add `: IDisposable`)
2. Quick action is empty and doesn't provide any action (Implement Interface...)
This is reproductible on other computers. I have tried the following:
   * VS repair
   * VS settings reset
   * Deletion of .vs folder
   * Deletion of all obj folders
Interestingly, the autompletion for adding a namespace contains 3 library names out of four. This seems to prove that the cache update gets interrupted, preventing the intellisense to work as expected.
I don't know what I can do (log) to provide more meaningful information. Please advise. The solution I'm working on cannot be shared outside my company.

**Expected Behavior**:
I'd expect intellisens to be working.

**Actual Behavior**:
More generally almost all intellisense features are broken:
   * Intellisense is not detecting any code issue (wrong method name...), only detecting when compiling
   * Not able to make missing using proposals
   * No quick action such as Extract method or such

**Comment by Pilchie** [comment]
What does the drop down at the top left of the file say?  Does it say "Miscellaneous files"?  If it does, are there any errors/warnings in your error list?  Can you try diagnosing a design-time build failure as described at https://github.com/dotnet/project-system/blob/master/docs/design-time-builds.md?

**Comment by frblondin** [comment]
The correct drop down list value is displayed. I could enable the design time logs and there is no error... despite the intellisense continues failing (no "Implement Interface..." action).

[xxx.Server.xxx_Compile_4bd3d94e.designtime.log.txt](https://github.com/dotnet/roslyn/files/1241054/xxx.Server.xxx_Compile_4bd3d94e.designtime.log.txt)
![failingintellisense](https://user-images.githubusercontent.com/5735134/29552160-64d68fea-8716-11e7-9597-6c417b3799c8.png)

**Comment by Pilchie** [comment]
Hmm - that is perplexing indeed.  @CyrusNajmabadi - any ideas here?

**Comment by frblondin** [comment]
Since Roslyn is open source, is there any way I could debug it?

**Comment by frblondin** [comment]
I was able to debug Roslyn and find some interesting stuff.

First my solution contains many projects but only two projects are really involved in the issue.

Flow of issues:
* [Executor.SemanticAnalysisEnabled(...)](https://github.com/dotnet/roslyn/blob/66284ff9b619bb9ae8182935b93162c88d51a53f/src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs#L481) is returning false. This prevents intellisense features to work correctly
* The reason is that [CompilationTracker.FinalizeCompilationAsync(...)](https://github.com/dotnet/roslyn/blob/56f79e51ce36af3b6e212f0c4354739f55d16a2b/src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs#L624) checks the status of each referenced project
* ... and asks for the metadataReference by calling [SolutionState.GetMetadataReferenceAsync(...)](https://github.com/dotnet/roslyn/blob/aeeec402a2f223580f36c298bbd9d92ffc94b330/src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs#L1734) which fails because [MetadataOnlyReference.GetOrBuildReference(...)](https://github.com/dotnet/roslyn/blob/175c1605927859bc3cfcd4a84b720b27c4054ba9/src/Workspaces/Core/Portable/Workspace/Solution/MetadataOnlyReference.cs#L34) returns a null reference. As image.IsEmpty returns true a null reference is returned.

So... I tried to understand why the image is empty for this referenced project. It is empty because the function [Compilation.FilterAndAppendDiagnostics(...)](https://github.com/dotnet/roslyn/blob/1c1fbc6bd3f22fa070b276b9cecb0c98ca03912a/src/Compilers/Core/Portable/Compilation/Compilation.cs#L1185) encounters some warnings (GetHashCode not implemented whereas Equals is overridden...) that are treated as errors, because TreatWarningsAsErrors is set to true. The problem is that a project rebuild doesn't show any warning. This Warning (that will be treated as an error) is ignored using a pragma. The problem is that there is another file for this type that completes the definition of X using a partial class. This other file doesn't use the pragmas. That is why in the end the warnings are simply ignored but Roslyn badly assumes that the project is in an error state.

There are two issues:
1. If a referenced project has warnings treated as errors this should not prevent the other projects depending on it to work
2. Warnings treated as errors are not displayed in VS or in the build result for my project whereas there are still active internally in the roslyn analysis results.

I tried to create a lightweight solution so you can reproduce the issue. I reproduced this simple organization where there are two projects, and the project (being referenced) having a partial class with pragma-ignored warnings. For some reason I couldn't reproduce the error and I can't spend too much time on this anylonger...

I hope that, thanks to the call flow description above, you will be able to better understand what can cause the problem.

**Comment by Pilchie** [comment]
Thanks for the debugging!  One thing you might try is enabling Full solution analysis from Tools/Options/Text Editor/C#/Advanced. It should hopefully show you the IDE errors for closed files as well.

@jasonmalinowski / @jcouv - looks like another skeleton assembly issue, maybe some part of the pragma stuff is skipped?  We should probably always ignore warnaserror?

**Comment by frblondin** [comment]
We are disabling Full solution analysis because of performance issues. Next time we are getting such errors we will try to enable it before debugging roslyn, hehe.

**Comment by jasonmalinowski** [comment]
Yeah, we should most definitely not be applying warn-as-error to skeleton assembly builds. That's...very not good. Awesome debugging job @frblondin!

I guess the fix is to amend [this line](https://github.com/dotnet/roslyn/blob/8dd264ba6b3147c82c8cc584bc8f07d9535291a9/src/Workspaces/Core/Portable/Workspace/Solution/MetadataOnlyImage.cs#L50), maybe? I admit I don't know what Emit does for warn-as-error here. Clearly the "Succeeded" is coming back false (which makes sense), but did it still do an emit to the stream anyways? Put another way, if we stop looking at Succeeded and actually look at the diagnostics to see if there were any 'real' errors, is that sufficient? Or do we have to turn of warn as error in the options before we even call emit? I'm hoping the former is the case, because otherwise that compilation for with new options might get more expensive, not sure.

@mavasani or @jaredpar or maybe @gafter, any idea what the contract is there?

**Comment by frblondin** [comment]
FYI I disabled all warn as errors but I'm still having the intellisense broken for other types/projects. The reason is that there are false positives in the diagnostics of type `error BC30002: Type 'IClassExplorer' is not defined`.

Having these false positives should be another issue that I will post separately, but this shows that creating the image even if there are errors (**even real errors, not only warn-as-error types**) for skeleton assembly is absolutely required. At this stage, I edit my type as if I was using a simple text editor without intellisense :-(

Can you possibly fix this issue in the current release (15.3), not in 15.5?

**Comment by jaredpar** [comment]
CC @jcouv

**Comment by CyrusNajmabadi** [comment]
Why are these errors being created?  Can you just fix the error?  The way skeletons work is that we should keep using the last successfully built skeleton until we can successfully make a new one.  If we can't even make one, then there's going to be problems.

**Comment by frblondin** [comment]
@CyrusNajmabadi the primary step is indeed to fix the errors when they are displayed. The problem is that there are some scenarios where false positives prevent Roslyn to successfully complete the analysis of a project B
- The issue I had where warn as errors were pragma-ignored in one file of a partial type, not another,
- I probably don't need to tell you that it happens that the Intellisense displays false positives and a compile through msbuild works. I experienced this many times and a quick [research](https://www.google.fr/search?q=intellisense+false+errors) confirms that this *happen*,
- Sometimes after switching between source control branches false positives are displayed.

Again, these are separate problems and should be addressed as other issues. What I don't understand is why Roslyn would on purpose implement a waterfall effect where a failure in Project B prevents from a simple class in Project A from using intellisense.

I don't know of all the ramifications but I imagine that the solution is to either change the skeleton assembly builds strategy OR to change the Executor.SemanticAnalysisEnabled(...) / CompilationTracker.FinalizeCompilationAsync(...) functions so that semantic analysis is still enabled if a referenced project has errors (being false positives or not...).

**Comment by CyrusNajmabadi** [comment]
> What I don't understand is why Roslyn would on purpose implement a waterfall effect where a failure in Project B prevents from a simple class in Project A from using intellisense.

There is nothing in intellisense that intentionally does this.  What's likely happening is that something deep in the bowels of Roslyn's workspace has broken, and it's causing everything to be busted.

**Comment by CyrusNajmabadi** [comment]
Note: IntelliSense is the term we use to refer to "completion lists".  I think you're complaining that lightbulbs and semantic diagnostics are not working.

That's actually intentional when we can't properly build a dependent project.  The reason for this (and @heejaechang can shed some more light on this), is that we need to prevent a massive cascading diagnostics problem.  i.e. in practice, when a dependent project doesn't build, user may often get literally hundreds of thousands of errors.  This is not helpful, and bogs down both the internal system, and also can overload the user experience.

So, when a dependent project fails to build, semantic diagnostics are suppressed.

**Comment by frblondin** [comment]
@CyrusNajmabadi you are right I'm more talking about semantic diagnostics.

I remember a few years ago when I was getting 9k+ errors (there was probably a 10k threashold) so I understand what you are referring to with the _cascading diagnostics problem_.

My main concern is that me, as a developer working on VS, could not use it correctly and my only way to understand how I could get VS working again was by debugging Roslyn. I imagine that if fixing all the scenarios that I listed in my previous message (and other scenarios I don't know about) is too complex the following changes would become mandatory:
1. Explain why semantic analysis is not possible
In the VS interface or in the VS Intellisense console a message of type `'Semantic analysis is temporarily disabled as a dependeant project XXX contains blocking errors'`. This would be so valuable compared to the cost of implementation I guess. What seemed to be a bug becomes a normal behavior as the developer has the beginning of the actions he needs to perform in order to get his solution working again.
2. Talking again about false errors in the VS analysis (whereas msbuild works): give the ability to reset everything (second bullet point from my list of scenarios above)
When I'm getting such errors I need to 1/ Stop VS, 2/ delete the .vs folder, and 3/ delete the obj folders in nested project directories, 4/ Start VS, 5/ Pray. Having a button that does a full reset for me would be much appreciated.

**Comment by CyrusNajmabadi** [comment]
> Explain why semantic analysis is not possible
In the VS interface or in the VS Intellisense console a message of type 'Semantic analysis is temporarily disabled as a dependeant project XXX contains blocking errors'. 

This is what should be happening.  @heejaechang Do you know why no message would be shown in this case?

**Comment by heejaechang** [comment]
of several causes, one of them report "design time build failures", but not all of them.

@Pilchie did we decide what to do here? are we going to show messages? or re-enable errors for open files?

**Comment by Pilchie** [comment]
Sorry for the delay - this came in when I was on vacation.  IIRC we decided that:

1. In the short term, we should introduce a new flag to handle the pre-restore case, and then bring back errors in open files for the case where we don't know that it's pre-restore.
2. In the medium term, we should make sure we collect and report errors from skeleton assembly parsing, even if full solution analysis is disabled, so that users can discover what the problem is.

**Comment by frblondin** [comment]
Thanks for the update and it seems that this would allow devs to get the errors - is the milestone still the same (15.5) or given the complexity of this fix it could be done sooner?

**Comment by heejaechang** [comment]
@Pilchie @jinujoseph @jcouv did we decide to support skeleton assembly with certain errors?

@davkean @Pilchie have we added events to indicate whether project system did everything to fully loaded projects?

**Comment by Pilchie** [comment]
> @davkean @Pilchie have we added events to indicate whether project system did everything to fully loaded projects?

See https://github.com/dotnet/roslyn/pull/24870

**Comment by heejaechang** [comment]
@Pilchie I am not talking about telemetry event but event like OnProjectFullyInitialized or something. @davkean said he will add one at some point. for 1 to make test less flaky (since we currently have no way to know when a solution is actually fully loaded) and 2 to make use to start show things like semantic errors.

**Comment by Pilchie** [comment]
Not yet - still in progress at https://github.com/dotnet/project-system/tree/features/LanguageService

**Comment by frblondin** [comment]
@Pilchie the link you shared is no longer working.

My solution is frequently not working... both with VS 2017 and VS 2019. My only workaround is to use funnel and only loads the projects I need to edit after a full compilation.

Please release a version with this fix - behind a feature toggling?

**Comment by jasonmalinowski** [comment]
@frblondin The link @Pilchie was to a branch that has now been merged and delivered into the current Visual Studio 2019 Previews.

@Pilchie Is there something new we should be consuming from that work? I admit the connection to this I'm a bit fuzzy on.

**Comment by Pilchie** [comment]
No, I don't think this API is done yet. Tracked by https://github.com/dotnet/project-system/issues/3427 and https://github.com/dotnet/project-system/issues/3425 AFAIK.

**Comment by frblondin** [comment]
It's been some time... would you have any update to share?

**Comment by CyrusNajmabadi** [comment]
Not planning on making changes to how we build compilatinos.
